From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mikhail Zinchenko <m1d3ry@proton.me>
Date: Tue, 1 Sep 2026 17:40:07 +0200
Subject: [PATCH] Add the generated xerces-c headers for a POSIX build

xerces-c has CMake probe the host and generate `config.h`, `Xerces_autoconf_config.hpp` and
`XercesVersion.hpp`. A build system that does not run CMake must supply all three.

`config.h` and `Xerces_autoconf_config.hpp` are hand-authored for a POSIX build, that is Linux glibc and
macOS, with the IconvGNU transcoder, the InMemory message loader, the POSIX file and mutex managers, no
ICU and no network accessor. `XMLCh` is `char16_t`. `XercesVersion.hpp` is the upstream template with its
four version numbers filled in, for xerces-c 3.2.1 at grammar serialization level 7.

Windows needs none of these, because it uses the system msxml6 parser instead of xerces.

A bump of the bundled xerces-c needs `XercesVersion.hpp` regenerated by hand. Do not send this commit
upstream.
---
 lib/xerces/src/config.h                       | 123 ++++++++++
 lib/xerces/src/xercesc/util/XercesVersion.hpp | 219 ++++++++++++++++++
 .../xercesc/util/Xerces_autoconf_config.hpp   |  84 +++++++
 3 files changed, 426 insertions(+)
 create mode 100644 lib/xerces/src/config.h
 create mode 100644 lib/xerces/src/xercesc/util/XercesVersion.hpp
 create mode 100644 lib/xerces/src/xercesc/util/Xerces_autoconf_config.hpp

diff --git a/lib/xerces/src/config.h b/lib/xerces/src/config.h
new file mode 100644
index 00000000..ba4ea265
--- /dev/null
+++ b/lib/xerces/src/config.h
@@ -0,0 +1,123 @@
+// Hand-authored replacement for the CMake-generated config.h of xerces-c (bundled under lib/xerces). Covers a
+// POSIX build (Linux glibc + macOS) with the IconvGNU transcoder, InMemory message loader, POSIX file/mutex
+// managers and no network accessor. Included by xerces sources under HAVE_CONFIG_H, which the build system sets.
+// Only macros relevant to the compiled subset are defined; the rest are left off.
+#ifndef MSIX_XERCES_CONFIG_H
+#define MSIX_XERCES_CONFIG_H
+
+// --- headers present on both Linux glibc and macOS ---
+#define HAVE_CTYPE_H 1
+#define HAVE_DLFCN_H 1
+#define HAVE_ERRNO_H 1
+#define HAVE_FCNTL_H 1
+#define HAVE_FLOAT_H 1
+#define HAVE_INTTYPES_H 1
+#define HAVE_LANGINFO_H 1
+#define HAVE_LIMITS_H 1
+#define HAVE_LOCALE_H 1
+#define HAVE_MEMORY_H 1
+#define HAVE_NETDB_H 1
+#define HAVE_NETINET_IN_H 1
+#define HAVE_STDBOOL_H 1
+#define HAVE_STDDEF_H 1
+#define HAVE_STDINT_H 1
+#define HAVE_STDIO_H 1
+#define HAVE_STDLIB_H 1
+#define HAVE_STRING_H 1
+#define HAVE_STRINGS_H 1
+#define HAVE_SYS_PARAM_H 1
+#define HAVE_SYS_SOCKET_H 1
+#define HAVE_SYS_STAT_H 1
+#define HAVE_SYS_TIME_H 1
+#define HAVE_SYS_TIMEB_H 1
+#define HAVE_SYS_TYPES_H 1
+#define HAVE_UNISTD_H 1
+#define HAVE_WCHAR_H 1
+#define HAVE_WCTYPE_H 1
+#define HAVE_ARPA_INET_H 1
+#define HAVE_ICONV_H 1
+
+// endian header differs between glibc and macOS/BSD
+#if defined(__APPLE__)
+#define HAVE_MACHINE_ENDIAN_H 1
+#else
+#define HAVE_ENDIAN_H 1
+#endif
+
+// --- functions ---
+#define HAVE_CLOCK_GETTIME 1
+#define HAVE_FTIME 1
+#define HAVE_GETADDRINFO 1
+#define HAVE_GETCWD 1
+#define HAVE_GETHOSTBYADDR 1
+#define HAVE_GETHOSTBYNAME 1
+#define HAVE_GETTIMEOFDAY 1
+#define HAVE_GMTIME_R 1
+#define HAVE_ICONV 1
+#define HAVE_ICONV_CLOSE 1
+#define HAVE_ICONV_OPEN 1
+#define HAVE_LOCALECONV 1
+#define HAVE_MBLEN 1
+#define HAVE_MBRLEN 1
+#define HAVE_MBSRTOWCS 1
+#define HAVE_MBSTOWCS 1
+#define HAVE_MEMMOVE 1
+#define HAVE_MEMSET 1
+#define HAVE_NL_LANGINFO 1
+#define HAVE_PATHCONF 1
+#define HAVE_PTHREAD 1
+#define HAVE_REALPATH 1
+#define HAVE_SETLOCALE 1
+#define HAVE_SNPRINTF 1
+#define HAVE_SOCKET 1
+#define HAVE_STRCASECMP 1
+#define HAVE_STRCHR 1
+#define HAVE_STRDUP 1
+#define HAVE_STRNCASECMP 1
+#define HAVE_STRRCHR 1
+#define HAVE_STRSTR 1
+#define HAVE_STRTOL 1
+#define HAVE_STRTOUL 1
+#define HAVE_TIMEGM 1
+#define HAVE_TOWLOWER 1
+#define HAVE_TOWUPPER 1
+#define HAVE_WCSRTOMBS 1
+#define HAVE_WCSTOMBS 1
+
+// --- language / library features ---
+#define HAVE_BOOL 1
+#define HAVE_CONST 1
+#define HAVE_CSTDINT 1
+#define HAVE_INLINE 1
+#define HAVE_NAMESPACES 1
+#define HAVE_STD_NAMESPACE 1
+#define HAVE_STD_LIBS 1
+#define HAVE_VOLATILE 1
+#define HAVE_PATH_MAX 1
+
+// Both glibc's iconv and macOS's system iconv take a non-const `char**`, so ICONV_USES_CONST_POINTER stays off.
+
+// --- sizes (LP64 on both Linux and macOS) ---
+#define SIZEOF_SHORT 2
+#define SIZEOF_INT 4
+#define SIZEOF_LONG 8
+#define SIZEOF_LONG_LONG 8
+#define SIZEOF_WCHAR_T 4
+
+// --- package identity ---
+#define PACKAGE "xerces-c"
+#define PACKAGE_NAME "xerces-c"
+#define PACKAGE_TARNAME "xerces-c"
+#define PACKAGE_VERSION "3.2.1"
+#define PACKAGE_STRING "xerces-c 3.2.1"
+#define PACKAGE_BUGREPORT ""
+#define PACKAGE_URL ""
+#define VERSION "3.2.1"
+
+// --- subsystem selection (POSIX, pack-only, no network / ICU) ---
+#define XERCES_USE_TRANSCODER_GNUICONV 1
+#define XERCES_USE_MSGLOADER_INMEMORY 1
+#define XERCES_USE_FILEMGR_POSIX 1
+#define XERCES_USE_MUTEXMGR_POSIX 1
+
+#endif
diff --git a/lib/xerces/src/xercesc/util/XercesVersion.hpp b/lib/xerces/src/xercesc/util/XercesVersion.hpp
new file mode 100644
index 00000000..66dc41dd
--- /dev/null
+++ b/lib/xerces/src/xercesc/util/XercesVersion.hpp
@@ -0,0 +1,219 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * $Id$
+ */
+
+#if !defined(XERCESC_INCLUDE_GUARD_XERCESVERSION_HPP)
+#define XERCESC_INCLUDE_GUARD_XERCESVERSION_HPP
+
+// ---------------------------------------------------------------------------
+// X E R C E S   V E R S I O N   H E A D E R   D O C U M E N T A T I O N
+
+/**
+ * User Documentation for Xerces Version Values:
+ *
+ *
+ *
+ * Xerces  Notes:
+ *
+ * Xerces Committers Documentation:
+ *
+ *    Xerces committers normally only need to modify one or two of the
+ *    following macros:
+ *
+ *      XERCES_VERSION_MAJOR
+ *      XERCES_VERSION_MINOR
+ *      XERCES_VERSION_REVISION
+ *
+ *    The integer values of these macros define the Xerces version number. All
+ *    other constants and preprocessor macros are automatically generated from
+ *    these three definitions.
+ *
+ *    The macro XERCES_GRAMMAR_SERIALIZATION_LEVEL has been added so that during
+ *    development if users are using the latest code they can use the grammar
+ *    serialization/deserialization features.  Whenever a change is made to the
+ *    serialization code this macro should be incremented.
+ *
+ * Xerces User Documentation:
+ *
+ *  The following sections in the user documentation have examples based upon
+ *  the following three version input values:
+ *
+ *    #define XERCES_VERSION_MAJOR 19
+ *    #define XERCES_VERSION_MINOR 3
+ *    #define XERCES_VERSION_REVISION 74
+ *
+ *  The minor and revision (patch level) numbers have two digits of resolution
+ *  which means that '3' becomes '03' in this example. This policy guarantees
+ *  that when using preprocessor macros, version 19.3.74 will be greater than
+ *  version 1.94.74 since the first will expand to 190374 and the second to
+ *  19474.
+ *
+ *  Preprocessor Macros:
+ *
+ *    _XERCES_VERSION defines the primary preprocessor macro that users will
+ *    introduce into their code to perform conditional compilation where the
+ *    version of Xerces is detected in order to enable or disable version
+ *    specific capabilities. The value of _XERCES_VERSION for the above example
+ *    will be 190374. To use it a user would perform an operation such as the
+ *    following:
+ *
+ *      #if _XERCES_VERSION >= 190374
+ *        // code specific to new version of Xerces...
+ *      #else
+ *        // old code here...
+ *      #endif
+ *
+ *    XERCES_FULLVERSIONSTR is a preprocessor macro that expands to a string
+ *    constant whose value, for the above example, will be "19_3_74".
+ *
+ *    XERCES_FULLVERSIONDOT is a preprocessor macro that expands to a string
+ *    constant whose value, for the above example, will be "19.3.74".
+ *
+ *    XERCES_VERSIONSTR is a preprocessor macro that expands to a string
+ *    constant whose value, for the above example, will be "19_3". This
+ *    particular macro is very dangerous if it were to be used for comparing
+ *    version numbers since ordering will not be guaranteed.
+ *
+ *    Xerces_DLLVersionStr is a preprocessor macro that expands to a string
+ *    constant whose value, for the above example, will be "19_3_74". This
+ *    macro is provided for backwards compatibility to pre-1.7 versions of
+ *    Xerces.
+ *
+ *  String Constants:
+ *
+ *    gXercesVersionStr is a global string constant whose value corresponds to
+ *    the value "19_3" for the above example.
+ *
+ *    gXercesFullVersionStr is a global string constant whose value corresponds
+ *    to the value "19_3_74" for the above example.
+ *
+ *  Numeric Constants:
+ *
+ *    gXercesMajVersion is a global integer constant whose value corresponds to
+ *    the major version number. For the above example its value will be 19.
+ *
+ *    gXercesMinVersion is a global integer constant whose value corresponds to
+ *    the minor version number. For the above example its value will be 3.
+ *
+ *    gXercesRevision is a global integer constant whose value corresponds to
+ *    the revision (patch) version number. For the above example its value will
+ *    be 74.
+ *
+ */
+
+// ---------------------------------------------------------------------------
+// X E R C E S   V E R S I O N   S P E C I F I C A T I O N
+
+/**
+ * MODIFY THESE NUMERIC VALUES TO COINCIDE WITH XERCES VERSION
+ * AND DO NOT MODIFY ANYTHING ELSE IN THIS VERSION HEADER FILE
+ */
+
+#define XERCES_VERSION_MAJOR 3
+#define XERCES_VERSION_MINOR 2
+#define XERCES_VERSION_REVISION 1
+
+/***
+ *
+ * XERCES_GRAMMAR_SERIALIZATION_LEVEL = 4   SchemaAttDef, SchemaElementDecl serialize fPSVIScope
+ * XERCES_GRAMMAR_SERIALIZATION_LEVEL = 5   XercesStep serializes the axis as an int
+ * XERCES_GRAMMAR_SERIALIZATION_LEVEL = 6   added fIsExternal to XMLEntityDecl
+ * XERCES_GRAMMAR_SERIALIZATION_LEVEL = 7   size of line/column fields has changed
+ *
+ ***/
+#define XERCES_GRAMMAR_SERIALIZATION_LEVEL 7
+
+/** DO NOT MODIFY BELOW THIS LINE */
+
+/**
+ * MAGIC THAT AUTOMATICALLY GENERATES THE FOLLOWING:
+ *
+ *	Xerces_DLLVersionStr, gXercesVersionStr, gXercesFullVersionStr,
+ *	gXercesMajVersion, gXercesMinVersion, gXercesRevision
+ */
+
+// ---------------------------------------------------------------------------
+// T W O   A R G U M E N T   C O N C A T E N A T I O N   M A C R O S
+
+// two argument concatenation routines
+#define CAT2_SEP_UNDERSCORE(a, b) #a "_" #b
+#define CAT2_SEP_PERIOD(a, b) #a "." #b
+#define CAT2_SEP_NIL(a, b) #a #b
+#define CAT2_RAW_NUMERIC(a, b) a ## b
+
+// two argument macro invokers
+#define INVK_CAT2_SEP_UNDERSCORE(a,b) CAT2_SEP_UNDERSCORE(a,b)
+#define INVK_CAT2_SEP_PERIOD(a,b)     CAT2_SEP_PERIOD(a,b)
+#define INVK_CAT2_STR_SEP_NIL(a,b)    CAT2_SEP_NIL(a,b)
+#define INVK_CAT2_RAW_NUMERIC(a,b)    CAT2_RAW_NUMERIC(a,b)
+
+// ---------------------------------------------------------------------------
+// T H R E E   A R G U M E N T   C O N C A T E N A T I O N   M A C R O S
+
+// three argument concatenation routines
+#define CAT3_SEP_UNDERSCORE(a, b, c) #a "_" #b "_" #c
+#define CAT3_SEP_PERIOD(a, b, c) #a "." #b "." #c
+#define CAT3_SEP_NIL(a, b, c) #a #b #c
+#define CAT3_RAW_NUMERIC(a, b, c) a ## b ## c
+#define CAT3_RAW_NUMERIC_SEP_UNDERSCORE(a, b, c) a ## _ ## b ## _ ## c
+
+// three argument macro invokers
+#define INVK_CAT3_SEP_UNDERSCORE(a,b,c) CAT3_SEP_UNDERSCORE(a,b,c)
+#define INVK_CAT3_SEP_PERIOD(a,b,c)     CAT3_SEP_PERIOD(a,b,c)
+#define INVK_CAT3_SEP_NIL(a,b,c)        CAT3_SEP_NIL(a,b,c)
+#define INVK_CAT3_RAW_NUMERIC(a,b,c)    CAT3_RAW_NUMERIC(a,b,c)
+#define INVK_CAT3_RAW_NUMERIC_SEP_UNDERSCORE(a,b,c)    CAT3_RAW_NUMERIC_SEP_UNDERSCORE(a,b,c)
+
+// ---------------------------------------------------------------------------
+// C A L C U L A T E   V E R S I O N   -   E X P A N D E D   F O R M
+
+#define MULTIPLY(factor,value) factor * value
+#define CALC_EXPANDED_FORM(a,b,c) ( MULTIPLY(10000,a) + MULTIPLY(100,b) + MULTIPLY(1,c) )
+
+// ---------------------------------------------------------------------------
+// X E R C E S   V E R S I O N   I N F O R M A T I O N
+
+// Xerces version strings; these particular macros cannot be used for
+// conditional compilation as they are not numeric constants
+
+#define XERCES_FULLVERSIONSTR INVK_CAT3_SEP_UNDERSCORE(XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR,XERCES_VERSION_REVISION)
+#define XERCES_FULLVERSIONDOT INVK_CAT3_SEP_PERIOD(XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR,XERCES_VERSION_REVISION)
+#define XERCES_FULLVERSIONNUM INVK_CAT3_SEP_NIL(XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR,XERCES_VERSION_REVISION)
+#define XERCES_VERSIONSTR     INVK_CAT2_SEP_UNDERSCORE(XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR)
+
+// Xerces C++ Namespace string, concatenated with full version string
+#define XERCES_PRODUCT xercesc
+#define XERCES_CPP_NAMESPACE INVK_CAT3_RAW_NUMERIC_SEP_UNDERSCORE(XERCES_PRODUCT,XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR)
+
+// original from Xerces header
+#define Xerces_DLLVersionStr XERCES_FULLVERSIONSTR
+
+const char* const    gXercesVersionStr = XERCES_VERSIONSTR;
+const char* const    gXercesFullVersionStr = XERCES_FULLVERSIONSTR;
+const unsigned int   gXercesMajVersion = XERCES_VERSION_MAJOR;
+const unsigned int   gXercesMinVersion = XERCES_VERSION_MINOR;
+const unsigned int   gXercesRevision   = XERCES_VERSION_REVISION;
+
+// Xerces version numeric constants that can be used for conditional
+// compilation purposes.
+
+#define _XERCES_VERSION CALC_EXPANDED_FORM (XERCES_VERSION_MAJOR,XERCES_VERSION_MINOR,XERCES_VERSION_REVISION)
+
+#endif // XERCESVERSION_HPP
diff --git a/lib/xerces/src/xercesc/util/Xerces_autoconf_config.hpp b/lib/xerces/src/xercesc/util/Xerces_autoconf_config.hpp
new file mode 100644
index 00000000..9e0ae66e
--- /dev/null
+++ b/lib/xerces/src/xercesc/util/Xerces_autoconf_config.hpp
@@ -0,0 +1,84 @@
+// Hand-authored replacement for the CMake-generated Xerces_autoconf_config.hpp (xerces-c bundled under
+// lib/xerces). Values are for a POSIX build (Linux glibc + macOS), matching the IconvGNU transcoder
+// configuration (XMLCh = char16_t). See also the companion config.h.
+#ifndef XERCES_AUTOCONFIG_CONFIG_HPP
+#define XERCES_AUTOCONFIG_CONFIG_HPP
+
+#define XERCES_AUTOCONF 1
+#define XERCES_HAVE_SYS_TYPES_H 1
+#define XERCES_HAVE_CSTDINT 1
+#define XERCES_HAVE_STDINT_H 1
+#define XERCES_HAVE_INTTYPES_H 1
+/* XERCES_HAVE_INTRIN_H — MSVC only */
+/* XERCES_HAVE_EMMINTRIN_H — x86 SSE2 only, left off for arm64/x86 portability */
+#define XERCES_INCLUDE_WCHAR_H 1
+
+#define XERCES_S16BIT_INT int16_t
+#define XERCES_S32BIT_INT int32_t
+#define XERCES_S64BIT_INT int64_t
+#define XERCES_U16BIT_INT uint16_t
+#define XERCES_U32BIT_INT uint32_t
+#define XERCES_U64BIT_INT uint64_t
+#define XERCES_XMLCH_T char16_t
+#define XERCES_SIZE_T size_t
+#define XERCES_SSIZE_T ssize_t
+
+#define XERCES_HAS_CPP_NAMESPACE 1
+#define XERCES_STD_NAMESPACE 1
+#define XERCES_NEW_IOSTREAMS 1
+/* XERCES_NO_NATIVE_BOOL — we have bool */
+/* XERCES_LSTRSUPPORT — no wide L"" literal support needed */
+/* XERCES_MFC_SUPPORT — Windows only */
+/* XERCES_HAVE_CPUID_INTRINSIC / SSE2_INTRINSIC / GETCPUID — left off (no SIMD) */
+/* XERCES_NO_MATCHING_DELETE_OPERATOR — not needed */
+/* XERCES_DLL_EXPORT — we build a static archive */
+#define XERCES_STATIC_LIBRARY 1
+#define XERCES_PLATFORM_EXPORT
+#define XERCES_PLATFORM_IMPORT
+#ifdef XERCES_DLL_EXPORT
+#  define DLL_EXPORT
+#endif
+
+// ---------------------------------------------------------------------------
+//  Include standard headers, if available, that we may rely on below.
+// ---------------------------------------------------------------------------
+#if defined(__cplusplus) && defined(HAVE_CSTDINT)
+#  include <cstdint>
+#endif
+#if XERCES_HAVE_STDINT_H
+#  include <stdint.h>
+#endif
+#if XERCES_HAVE_INTTYPES_H
+#  include <inttypes.h>
+#endif
+#if XERCES_HAVE_SYS_TYPES_H
+#  include <sys/types.h>
+#endif
+#if XERCES_INCLUDE_WCHAR_H
+#  include <wchar.h>
+#endif
+
+typedef XERCES_SIZE_T				XMLSize_t;
+typedef XERCES_SSIZE_T				XMLSSize_t;
+
+#define XERCES_SIZE_MAX SIZE_MAX
+#define XERCES_SSIZE_MAX PTRDIFF_MAX
+
+typedef XERCES_XMLCH_T				XMLCh;
+
+typedef XERCES_U16BIT_INT			XMLUInt16;
+typedef XERCES_U32BIT_INT			XMLUInt32;
+typedef XERCES_U64BIT_INT			XMLUInt64;
+
+typedef XERCES_S16BIT_INT			XMLInt16;
+typedef XERCES_S32BIT_INT			XMLInt32;
+typedef XERCES_S64BIT_INT			XMLInt64;
+
+typedef XMLUInt64			        XMLFilePos;
+typedef XMLUInt64			        XMLFileLoc;
+
+#if defined(_DEBUG)
+#define XERCES_DEBUG
+#endif
+
+#endif
